docs: refresh Public Interface & align how-to guides for Airflow 3.0+ (#52297)
mistercrunch/Airflow: airflow-core/docs/public-airflow-interface.rst

    specific language governing permissions and limitations
    under the License.

**PUBLIC INTERFACE FOR AIRFLOW 3.0+**
=====================================

.. warning::

   **This documentation covers the Public Interface for Airflow 3.0+**

   If you are using Airflow 2.x, please refer to the
   `Airflow 2.11 Public Interface Documentation <https://airflow.apache.org/docs/apache-airflow/2.11.0/public-airflow-interface.html>`_

>    for the legacy interface.

Reviewer: cc @potiuk is this in line with what you expected?

Public Interface can be useful for building custom tools and integrations with other systems,
and for automating certain aspects of the Airflow workflow.

> The primary public interface for DAG Authors and task execution is using task SDK

Reviewer: ```suggestion
The primary public interface for DAG Authors and task execution is using task SDK
Airflow task SDK is the primary public interface for DAG Authors and for task execution
```

> In Airflow 3.0+, the primary public interface for DAG authors and task execution is the

Reviewer: ```suggestion
The primary public interface for DAG authors and task execution is the
```

The docs are versioned so no need for this.

Author: Yes removed all occurance of 3.0

> :doc:`airflow.sdk namespace <core-concepts/taskflow>`. Direct access to the metadata database

Reviewer: ```suggestion
using task SDK :doc:`airflow.sdk namespace <core-concepts/taskflow>`. Direct access to the metadata database
```

Give or take on the syntax

Author: Added it

from task code is no longer allowed. Instead, use the :doc:`Stable REST API <stable-rest-api-ref>`,

> `Python Client <https://github.com/apache/airflow-client-python>`_, or Task Context methods.

Reviewer: ```suggestion
`Python Client <https://github.com/apache/airflow-client-python>`_, or Task Context methods.
```
`Task Context methods.` is not v clear to me.

Author: Changed it

Using the Public Interface for DAG Authors
==========================================

> The primary interface for DAG authors in Airflow 3.0+ is the :doc:`airflow.sdk namespace <core-concepts/taskflow>`.

Reviewer: ```suggestion
The primary interface for DAG Authors is :doc:`airflow.sdk namespace <core-concepts/taskflow>`.
```

Author: Changed it

This provides a stable, well-defined interface for creating DAGs and tasks that is not subject to internal
implementation changes. The goal of this change is to decouple DAG authoring from Airflow internals (Scheduler,

> API Server, etc.), providing a forward-compatible, stable interface for writing and maintaining DAGs across Airflow versions.

Reviewer: ```suggestion
API Server, etc.), providing a version-agnostic, stable interface for writing and maintaining DAGs across Airflow versions.
```

Cannot commit to fwd compatible as of now i guess. CC @ashb / @kaxil

Author: Thats right

**Key Imports from airflow.sdk:**

**Classes:**

* ``Asset``
* ``BaseNotifier``
* ``BaseOperator``
* ``BaseOperatorLink``
* ``BaseSensorOperator``
* ``Connection``
* ``Context``
* ``DAG``
* ``EdgeModifier``
* ``Label``
* ``ObjectStoragePath``
* ``Param``
* ``TaskGroup``

> * ``Variable``

Reviewer: `BaseHook` merged in main too.

Author: Added it now

==========================================

The primary interface for DAG Authors is the :doc:`airflow.sdk namespace <core-concepts/taskflow>`.
This provides a stable, well-defined interface for creating DAGs and tasks that is not subject to internal
implementation changes. The goal of this change is to decouple DAG authoring from Airflow internals (Scheduler,
API Server, etc.), providing a version-agnostic, stable interface for writing and maintaining DAGs across Airflow versions.

**Key Imports from airflow.sdk:**

**Classes:**

* ``Asset``
* ``BaseHook``
* ``BaseNotifier``
* ``BaseOperator``
* ``BaseOperatorLink``
* ``BaseSensorOperator``
* ``Connection``
* ``Context``
* ``DAG``
* ``EdgeModifier``
* ``Label``
* ``ObjectStoragePath``
* ``Param``
* ``TaskGroup``
* ``Variable``

**Decorators and Functions:**

* ``@asset``
* ``@dag``
* ``@setup``
* ``@task``
* ``@task_group``
* ``@teardown``
* ``chain``
* ``chain_linear``
* ``cross_downstream``
* ``get_current_context``
* ``get_parsing_context``

**Migration Example:**

.. code-block:: python

    # Old (Airflow 2.x)
    from airflow.models import DAG
    from airflow.decorators import task

    # New (Airflow 3.x)

>     from airflow.sdk import DAG, task

Reviewer 2: Why do we need migration instructions here?
We should forward users to the migration doc 
https://airflow.apache.org/docs/apache-airflow/stable/installation/upgrading_to_airflow3.html

We should avoid having split content across docs

Author: removed the duplicates @eladkal

   airflow.sdk namespace instead.

.. note::
   The :class:`~airflow.models.dagrun.DagRun` class is used internally by Airflow for DAG run
   management. DAG authors should access DAG run information through the Task Context via
   :func:`~airflow.sdk.get_current_context` or use the :class:`~airflow.sdk.types.DagRunProtocol`
   interface.

.. _pythonapi:operators:

Operators
=========

The base classes :class:`~airflow.sdk.BaseOperator` and :class:`~airflow.sdk.BaseSensorOperator` are public and may be extended to make new operators.

The recommended base class for new operators is :class:`~airflow.sdk.BaseOperator`
from the airflow.sdk namespace.

Subclasses of BaseOperator which are published in Apache Airflow are public in *behavior* but not in *structure*.  That is to say, the Operator's parameters and behavior is governed by semver but the methods are subject to change at any time.

Task Instances
==============

Task instances are the individual runs of a single task in a DAG (in a DAG Run). They are available in the context
passed to the execute method of the operators via the :class:`~airflow.sdk.types.RuntimeTaskInstanceProtocol` class.

> In Airflow 3.0+, task instances are accessed through the Task Context via :func:`~airflow.sdk.get_current_context`

Reviewer: Don't add versions in docs please. Remove all such occurrences

Author: Removed it

References for the modules used in dags are here:

.. note::
   The airflow.sdk namespace provides the primary interface for DAG Authors.
   For detailed API documentation, see the `Task SDK Reference <https://airflow.apache.org/docs/task-sdk/stable/>`_.

.. note::
   The :class:`~airflow.models.dagbag.DagBag` class is used internally by Airflow for loading DAGs
   from files and folders. DAG Authors should use the :class:`~airflow.sdk.DAG` class from the
   airflow.sdk namespace instead.

.. note::
   The :class:`~airflow.models.dagrun.DagRun` class is used internally by Airflow for DAG run
   management. DAG Authors should access DAG run information through the Task Context via
   :func:`~airflow.sdk.get_current_context` or use the :class:`~airflow.sdk.types.DagRunProtocol`
   interface.

.. _pythonapi:operators:

Operators
=========

The base classes :class:`~airflow.sdk.BaseOperator` and :class:`~airflow.sdk.BaseSensorOperator` are public and may be extended to make new operators.

The recommended base class for new operators is :class:`~airflow.sdk.BaseOperator`

> from the airflow.sdk namespace.

Reviewer: Makes it sound like there's another option, maybe get rid of this bit?

Author: Changed it

Subclasses of BaseOperator which are published in Apache Airflow are public in *behavior* but not in *structure*.  That is to say, the Operator's parameters and behavior is governed by semver but the methods are subject to change at any time.

Task Instances
==============

Task instances are the individual runs of a single task in a DAG (in a DAG Run). They are available in the context

> passed to the execute method of the operators via the :class:`~airflow.sdk.types.RuntimeTaskInstanceProtocol` class.

Reviewer: Hmmm, do we want to expose the `RuntimeTaskInstanceProtocol` in docs?
cc @kaxil

Author: I have removed it for now as it is not exposed.

the stable interface for task instance operations.

.. note::
   Task Context and RuntimeTaskInstanceProtocol are part of the airflow.sdk namespace.
   For detailed API documentation, see the `Task SDK Reference <https://airflow.apache.org/docs/task-sdk/stable/>`_.

Task Instance Keys
==================

Task instance keys are unique identifiers of task instances in a DAG (in a DAG Run). A key is a tuple that consists of
``dag_id``, ``task_id``, ``run_id``, ``try_number``, and ``map_index``.

Direct access to task instance keys via the :class:`~airflow.models.taskinstance.TaskInstance`
model is no longer allowed from task code. Instead, use the Task Context via :func:`~airflow.sdk.get_current_context`
to access task instance information.

Example of accessing task instance information through Task Context:

.. code-block:: python

    from airflow.sdk import get_current_context

    def my_task():
        context = get_current_context()
        ti = context["ti"]

        dag_id = ti.dag_id
        task_id = ti.task_id
        run_id = ti.run_id
        try_number = ti.try_number
        map_index = ti.map_index

>         print(f"Task: {dag_id}.{task_id}, Run: {run_id}, Try: {try_number}, Map: {map_index}")

Reviewer: ```suggestion
        print(f"Task: {dag_id}.{task_id}, Run: {run_id}, Try: {try_number}, Map Index: {map_index}")
```

  _api/airflow/hooks/index

Public Airflow utilities
========================

When writing or extending Hooks and Operators, DAG Authors and developers can
use the following classes:

* The :class:`~airflow.sdk.Connection`, which provides access to external service credentials and configuration.
* The :class:`~airflow.sdk.Variable`, which provides access to Airflow configuration variables.
* The :class:`~airflow.sdk.execution_time.xcom.XCom` which are used to access to inter-task communication data.

Connection and Variable operations should be performed through the Task Context using
:func:`~airflow.sdk.get_current_context` and the task instance's methods, or through the airflow.sdk namespace.
Direct database access to :class:`~airflow.models.connection.Connection` and :class:`~airflow.models.variable.Variable`
models is no longer allowed from task code.

> Example of accessing connections and variables through Task Context:

Reviewer: ```suggestion
Example of accessing Connections and Variables through Task Context:
```

.. code-block:: python

    from airflow.sdk import get_current_context

    def my_task():
        context = get_current_context()

        conn = context["conn"]
        my_connection = conn.get_connection("my_connection_id")

        var = context["var"]

>         my_variable = var.get("my_variable_name")

Reviewer: Have you validated this?
The connection certainly doesn't look right..

Author: Changed it.

Example of using airflow.sdk namespace directly:

.. code-block:: python

    from airflow.sdk import Connection, Variable

>     conn = Connection.get_connection("my_connection_id")

Reviewer: ```suggestion
    conn = Connection.get("my_connection_id")
```

This isn't right. its

Example of using Task Context instead of direct database access:

.. code-block:: python

    from airflow.sdk import dag, get_current_context, task
    from airflow.utils.state import DagRunState
    from datetime import datetime

    @dag(dag_id="example_dag", start_date=datetime(2025, 1, 1), schedule="@hourly", tags=["misc"], catchup=False)
    def example_dag():

        @task(task_id="check_dagrun_state")
        def check_state():
            context = get_current_context()
            ti = context["ti"]
            dag_run = context["dag_run"]

            # Use Task Context methods instead of direct DB access
            dr_count = ti.get_dr_count(dag_id="example_dag")
            dagrun_state = ti.get_dagrun_state(dag_id="example_dag", run_id=dag_run.run_id)

            return f"DAG run count: {dr_count}, current state: {dagrun_state}"

        check_state()

    example_dag()

.. note::

   **For Airflow 2.x users**: If you are using Airflow 2.x, please refer to the
   `Airflow 2.11 Public Interface Documentation <https://airflow.apache.org/docs/apache-airflow/2.11.0/public-airflow-interface.html>`_

>    for the legacy interface.

Reviewer: Is this needed again? Feels repetitive

Author: I have removed it but then I had added this link for user who has reached the end of the page and move to previous version of documentation.
